### ISSUE TITLE
MacOSX / clang linking errors

### ISSUE THREAD

**Issue by russgayle** [embree/embree#122]
I'm having trouble linking to embree from a project we're working on. Briefly: 
* We include embree as a submodule on the project and build it along with our build (rather than installing it on the system).
* We alter the default configuration to use `INTERNAL` for the tasking and disable ISPC (to try to be uniform on other platforms where we're not using TBB or ISPC for now). 
* We also set `EMBREE_STATIC_LIB` as OFF

Embree builds properly as far as I can tell, but our build fails at the linking stage when we try to link against it. I've also verified that I can build and run tutorials on my machine as well, so I'm thinking it's something with how I'm linking it into our project. The associated `TARGET_LINK_LIBRARIES` command in our `CMakeLists.txt` include `embree simd tasking sys lexers`. The command-line used during linking includes:
```
3rdparty/embree/libembree.2.15.0.dylib 3rdparty/embree/libsimd.a 3rdparty/embree/libtasking.a 3rdparty/embree/libsys.a 3rdparty/embree/liblexers.a 3rdparty/embree/libembree_sse42.a 3rdparty/embree/libembree_avx.a 3rdparty/embree/libembree_avx2.a
```

where `3rdparty/embree` is the path to where the embree submodule is built within the build-tree, i.e. it's included like:

```
add_subdirectory(${ROOT_DIRECTORY}/3rdparty/embree ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/embree)
```

The message it returns is: 
```
Undefined symbols for architecture x86_64:
  "embree::State::verbosity(unsigned long)", referenced from:
      embree::BVHN<8>::preBuild(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libembree_avx.a(bvh.cpp.avx.cpp.o)
      embree::BVHN<8>::postBuild(double) in libembree_avx.a(bvh.cpp.avx.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

As far as I can tell, that symbol isn't exported in any of the `*.dylib` nor `*.a` files though I see it's defined and compiled in `kernels/common/state.cpp` (and, there's an associated `state.cpp.o`). (Checked by using `nm -gU`.) Any thoughts as to why I'm getting this or what I can do to remedy it?

If it helps, I'm running MacOSX (10.11.6, El Capitan) and using llvm/clang to build: 
```
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Lastly, the same setup does seem to work properly on our Windows build machine, so it seems specific to MacOSX and/or clang. 

Any help would be appreciated! Thanks!

**Comment by svenwoop** [comment]
Just created a little reproducer for your issue, but I see no problems. The following works properly on my Mac:

CMakeLists.txt:
===============

PROJECT(use_embree)

CMAKE_MINIMUM_REQUIRED(VERSION 2.8.11)

OPTION(EMBREE_STATIC_LIB OFF)
OPTION(EMBREE_ISPC_SUPPORT OFF)
SET(EMBREE_TASKING_SYSTEM "INTERNAL" CACHE STRING "Selects tasking system")

ADD_SUBDIRECTORY(embree)

INCLUDE_DIRECTORIES(${PROJECT_SOURCE_DIR}/embree/include)
ADD_EXECUTABLE(use_embree use_embree.cpp)
TARGET_LINK_LIBRARIES(use_embree embree)

use_embree.cpp
===============

#include "embree2/rtcore.h"

#include <xmmintrin.h>
//#include <pmmintrin.h> // use this to get _MM_SET_DENORMALS_ZERO_MODE when compiling for SSE3 or higher

#if !defined(_MM_SET_DENORMALS_ZERO_MODE)
#define _MM_DENORMALS_ZERO_ON   (0x0040)
#define _MM_DENORMALS_ZERO_OFF  (0x0000)
#define _MM_DENORMALS_ZERO_MASK (0x0040)
#define _MM_SET_DENORMALS_ZERO_MODE(x) (_mm_setcsr((_mm_getcsr() & ~_MM_DENORMALS_ZERO_MASK) | (x)))
#endif

int main(int argc, char* argv[])
{
  /* for best performance set FTZ and DAZ flags in MXCSR control and status register */
  _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);
  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);

  /* create new Embree device */
  RTCDevice device = rtcNewDevice("verbose=1");

  /* ddelete device again */
  rtcDeleteDevice(device);
  
  return 0;
}

Could you please verify if this works on your machine.

**Comment by russgayle** [comment]
This worked, and yeah, I'd tried something similar to that too and it worked. The main difference was I linked embree to a shared library and then linked that library to an executable (closer to how our project is linking embree). I was hoping you guys might've had some build insight that I was missing thus far :)

Anyway, I'll probably spend some more time today trying to resolve this. I'll let you know what I find out! Thanks!

**Comment by russgayle** [comment]
Hey! I dug into this a bit more today. There's something bizarre going on in our build between shared/static libraries and embree in one particular library of ours. I did two main things:
* Made a test subproject within our build, similar to source above but built with a similar configuration to the problematic one. That worked (built, linked, ran). 
* On the problematic library, I stripped out most of code in the to get to a point where I could build and link. I did get to a working state and slowed added things back. But, that lead down a complicated path; the end result was a strange state where including a `std::cerr << std::endl;` caused the problem above. Looking at the linker, there's something bizarre going on with the stdlibc++ I think. 

Anyway, this looks to be more to due our code than embree. I did manage to get my library into a good-enough state by building embree into a static library on Mac (and, shared on Windows). That'll do for us for now, I'll close this issue. 

Thanks for your help!
